File: NSIDC/AMSR_E_L3_SeaIce6km_V15_20050118.hdf.s.py
"""
Copyright (C) 2014 John Evans

This example code illustrates how to access and visualize an NSIDC AMSR-E L3 
SeaIce 6km HDF-EOS2 Grid file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python AMSR_E_L3_SeaIce6km_V15_20050118.hdf.s.py

The HDF file must be in your current working directory.

Tested under: Python 3.9.1 :: Miniconda
Last updated: 2022-02-04
"""

import os
import re
import matplotlib as mpl
import matplotlib.pyplot as plt
from mpl_toolkits.basemap import Basemap
import pyproj
import numpy as np

USE_GDAL = False

def run(FILE_NAME):

   # Identify the data field.    
    DATAFIELD_NAME = 'SI_06km_SH_89V_DAY'

    if USE_GDAL:    
        import gdal
        GRID_NAME = 'SpPolarGrid06km'
    
        gname = 'HDF4_EOS:EOS_GRID:"{0}":{1}:{2}'.format(FILE_NAME,
                                                         GRID_NAME,
                                                         DATAFIELD_NAME)
        gdset = gdal.Open(gname)
        data = gdset.ReadAsArray().astype(np.float64)

        # Read projection parameters from global attribute.
        meta = gdset.GetMetadata()
        x0, xinc, _, y0, _, yinc = gdset.GetGeoTransform()
        nx, ny = (gdset.RasterXSize, gdset.RasterYSize)
        del gdset
    else:
        from pyhdf.SD import SD, SDC
        hdf = SD(FILE_NAME, SDC.READ)

        # Read dataset.
        data2D = hdf.select(DATAFIELD_NAME)
        data = data2D[:,:].astype(np.float64)

    # There are multiple girds in this file.
    # Thus, simple regular expression search for 
    # UpperLeft/LowerRightPoint from StructMetadata.0 won't work. 
    # 
    # Use HDFView and look for the following parameters:
    #
    # GROUP=GRID_2
    #       GridName="SpPolarGrid06km"
    #	XDim=1264
    #	YDim=1328
    #	UpperLeftPointMtrs=(-3950000.000000,4350000.000000)
    #	LowerRightMtrs=(3950000.000000,-3950000.000000)
    ny, nx = data.shape
    x1 = 3950000
    x0 = -3950000
    y0 = 4350000
    y1 = -3950000
    xinc = (x1 - x0) / nx
    yinc = (y1 - y0) / ny
        
    # Apply the attributes information [1].
    data[data == 0] = np.nan
    data *= 0.1
    data = np.ma.masked_array(data, np.isnan(data))

    x = np.linspace(x0, x0 + xinc*nx, nx)
    y = np.linspace(y0, y0 + yinc*ny, ny)
    xv, yv = np.meshgrid(x, y)
    args = ["+proj=stere",
            "+lat_0=-90",
            "+lon_0=0",
            "+lat_ts=-70",
            "+k=1",
            "+es=0.006693883",
            "+a=6378273",
            "+x_0=0",
            "+y_0=0",
            "+ellps=WGS84",
            "+datum=WGS84"]
    pstereo = pyproj.Proj(' '.join(args))
    wgs84 = pyproj.Proj("+init=EPSG:4326") 
    lon, lat= pyproj.transform(pstereo, wgs84, xv, yv)
    units = 'K'
    long_name = DATAFIELD_NAME

    m = Basemap(projection='spstere', resolution='l', boundinglat=-45, lon_0 = 0)
    m.drawcoastlines(linewidth=0.5)
    m.drawparallels(np.arange(0, 91, 20), labels=[1, 0, 0, 0])
    m.drawmeridians(np.arange(-180, 181, 45), labels=[0, 0, 0, 1])
    m.pcolormesh(lon, lat, data, latlon=True)
    cb = m.colorbar()
    cb.set_label(units)

    basename = os.path.basename(FILE_NAME)
    plt.title('{0}\n{1}'.format(basename, long_name))
    fig = plt.gcf()
    pngfile = "{0}.s.py.png".format(basename)
    fig.savefig(pngfile)



if __name__ == "__main__":

    hdffile = 'AMSR_E_L3_SeaIce6km_V15_20050118.hdf'
    run(hdffile)
    
# Reference
# [1] https://nsidc.org/data/AE_SI6/versions/3

File: NSIDC/AMSR_E_L3_SeaIce6km_V15_20050118_hdf_s.m
%
%  This example code illustrates how to access and visualize NSIDC
%  AMSR_E Sea Ice 6km v15 L3 HDF-EOS2 Grid file in MATLAB.
%
%  If you have any questions, suggestions, comments on this example, 
% please use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS
% data product that is not listed in the HDF-EOS  Comprehensive Examples 
% page (http://hdfeos.org/zoo), feel free to contact us at 
% eoshelp@hdfgroup.org or post it at the HDF-EOS Forum 
% (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
%  $matlab -nosplash -nodesktop -r AMSR_E_L3_SeaIce6km_V15_20050118_hdf_s
%
% Tested under: MATLAB R2021a
% Last updated: 2022-02-04

import matlab.io.hdfeos.*

% Opening the HDF-EOS2 Grid File
FILE_NAME = 'AMSR_E_L3_SeaIce6km_V15_20050118.hdf';
file_id = gd.open(FILE_NAME, 'rdonly');

% Reading Data from a Data Field
GRID_NAME = 'SpPolarGrid06km';
grid_id = gd.attach(file_id, GRID_NAME);

DATAFIELD_NAME = 'SI_06km_SH_89V_DAY';
data1 = gd.readField(grid_id, DATAFIELD_NAME);

% Convert the data to double type for plot
data = double(data1);

% Get information about the spatial extents of the grid.
% [xdimsize, ydimsize, upleft, lowright] = gd.gridInfo(grid_id);


% Detach Grid object.
gd.detach(grid_id);
gd.close(file_id);

% Read two binary files for lat/lon.
latfile = fopen('pss06lats_v3.dat','r');
lonfile = fopen('pss06lons_v3.dat','r');

% Read the binary files as 4 byte integer to an array of 1792 x 1216.
l1 = fread(latfile, [1328 1264], 'int');
l2 = fread(lonfile, [1328 1264], 'int');

% Divide the values by 100,000 to get lat/lon.
lat = l1/100000;
lon = l2/100000;

% Handle fill value.
data(data==0) = NaN;

scale = 0.1;

% Apply scale factor.
data = data*scale;

% Plot the data using contourfm and axesm
latlim = [floor(min(min(lat))),ceil(max(max(lat)))];
lonlim = [floor(min(min(lon))),ceil(max(max(lon)))];
min_data = floor(min(min(data)));
max_data = ceil(max(max(data)));

f = figure('Name', FILE_NAME, 'Renderer', 'zbuffer', ...
           'Position', [0,0,800,600], 'visible', 'off');
pole=[-90 0 0];
axesm('MapProjection','stereo','MapLatLimit',latlim,'MapLonLimit',lonlim, ...
      'Origin', pole ,'Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel','north')
coast = load('coastlines.mat');
scatterm(lat(:), lon(:), 1, data(:));
colormap('Jet');
h = colorbar();
plotm(coast.coastlat,coast.coastlon,'k')
title({FILE_NAME; DATAFIELD_NAME}, 'Interpreter', 'None', ...
      'FontSize',12,'FontWeight','bold');
units = 'K';
set (get(h, 'title'), 'string', units, 'FontSize',12,'FontWeight','bold');
saveas(f,[FILE_NAME '.s.m.png']);
exit;

